src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,244 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "ObjectCacheStore.h"
+#include "Utils.h"
+#include <experimental/filesystem>
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_immutable_obj_cache
+#undef dout_prefix
+#define dout_prefix *_dout << "ceph::cache::ObjectCacheStore: " << this << " " \
+                           << __func__ << ": "
+
+namespace efs = std::experimental::filesystem;
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+ObjectCacheStore::ObjectCacheStore(CephContext *cct, ContextWQ* work_queue)
+      : m_cct(cct), m_work_queue(work_queue), m_rados(new librados::Rados()),
+        m_ioctxs_lock("ceph::cache::ObjectCacheStore::m_ioctxs_lock") {
+
+  object_cache_max_size =
+    m_cct->_conf.get_val<Option::size_t>("immutable_object_cache_max_size");
+
+  std::string cache_path = m_cct->_conf.get_val<std::string>("immutable_object_cache_path");
+  m_cache_root_dir = cache_path + "/ceph_immutable_obj_cache/";
+
+  //TODO(): allow to set cache level
+  m_policy = new SimplePolicy(m_cct, object_cache_max_size/(4096*1024), 0.1);
+}
+
+ObjectCacheStore::~ObjectCacheStore() {
+  delete m_policy;
+}
+
+int ObjectCacheStore::init(bool reset) {
+  ldout(m_cct, 20) << dendl;
+
+  int ret = m_rados->init_with_context(m_cct);
+  if(ret < 0) {
+    lderr(m_cct) << "fail to init Ceph context" << dendl;
+    return ret;
+  }
+
+  ret = m_rados->connect();
+  if(ret < 0 ) {
+    lderr(m_cct) << "fail to conect to cluster" << dendl;
+    return ret;
+  }
+
+  //TODO(): fsck and reuse existing cache objects
+  if (reset) {
+    if (efs::exists(m_cache_root_dir)) {
+      efs::remove_all(m_cache_root_dir);
+    }
+    efs::create_directories(m_cache_root_dir);
+  }
+
+  return ret;
+}
+
+int ObjectCacheStore::shutdown() {
+  ldout(m_cct, 20) << dendl;
+
+  m_rados->shutdown();
+  return 0;
+}
+
+int ObjectCacheStore::init_cache() {
+  ldout(m_cct, 20) << dendl;
+  std::string cache_dir = m_cache_root_dir;
+
+  int dir = m_dir_num - 1;
+  while (dir >= 0) {
+    efs::create_directories(cache_dir + "/" + std::to_string(dir));
+    dir --;
+  }
+  return 0;
+}
+
+int ObjectCacheStore::do_promote(std::string pool_name, std::string object_name) {
+  ldout(m_cct, 20) << "to promote object = "
+                   << object_name << " from pool: "
+                   << pool_name << dendl;
+
+  int ret = 0;
+  std::string cache_file_name =  pool_name + object_name;
+  {
+    Mutex::Locker _locker(m_ioctxs_lock);
+    if (m_ioctxs.find(pool_name) == m_ioctxs.end()) {
+      librados::IoCtx* io_ctx = new librados::IoCtx();
+      ret = m_rados->ioctx_create(pool_name.c_str(), *io_ctx);
+      if (ret < 0) {
+        lderr(m_cct) << "fail to create ioctx" << dendl;
+        return ret;
+      }
+      m_ioctxs.emplace(pool_name, io_ctx);
+    }
+  }
+
+  assert(m_ioctxs.find(pool_name) != m_ioctxs.end());
```

Comment:
Use `ceph_assert` instead of `assert`, here and in all other cases?

Reply:
done